rosco-core/src/main/groovy/com/netflix/spinnaker/rosco/providers/util/ImageNameFactory.groovy

```diff
@@ -16,27 +16,73 @@
-interface ImageNameFactory {
+import java.time.Clock
 
   /**
-   * This method is responsible for:
-   *   - Producing an image name.
-   *   - (If one or more package names are specified) Attempting to derive the AppVersion descriptor from the
-   *     first package name. If the AppVersion could not be derived, this value will be null.
-   *
-   * This method always returns a list of size 3 with the following elements:
-   *   1) A derived image name (to be used for naming the image being baked).
-   *   2) The appversion string (to be used for tagging the newly-baked image).
-   *   3) The updated list of packages to be used for overriding the passed package list (this will be removed once the
-   *      temporary workaround described above is removed).
-   *
-   * This function is not required to return the same image name on multiple invocations with the same bake request.
-   *
-   * Returns [imageName, appVersionStr, packagesParameter].
+   * Default implementation of ImageNameFactory considers only package_name, a timestamp, and base_os.
+   * For more fine grained conventions, extend this class, override the required methods and alter your
+   * provider specific RoscoConfiguration to reflect the change.
    */
-  def deriveImageNameAndAppVersion(BakeRequest bakeRequest, BakeOptions.Selected selectedOptions)
+
+@Slf4j
+public class ImageNameFactory {
+
+  @Autowired
+  Clock clock
+
+  /**
+   * Attempts to produce an appVersionStr from the first packageName; to be used for tagging the newly-baked image
+   */
+  def buildAppVersionStr(BakeRequest bakeRequest, PackageNameConverter.OsPackageName osPackageName) {
+    String appVersionStr = PackageNameConverter.buildAppVersionStr(bakeRequest, osPackageName)
+
+    if (!AppVersion.parseName(appVersionStr)) {
+      if (appVersionStr) {
+        log.debug("AppVersion.parseName() was unable to parse appVersionStr=$appVersionStr.")
+      }
+      return null
+    }
+    return appVersionStr
+  }
+
+  /**
+   * Produces an imageName either from the BakeRequest.ami_name or the first package to be installed.
+   * This is to be used for naming the image being baked. Note that this function is not required to
+   * return the same image name on multiple invocations with the same bake request
+   */
+  def buildImageName(BakeRequest bakeRequest, List<PackageNameConverter.OsPackageName> osPackageNames) {
+    String timestamp = clock.millis()
+    def emptyBasePackage = new PackageNameConverter.OsPackageName()
+    emptyBasePackage.with {
+      name = ""
+      arch = "all"
+    }
+
+    PackageNameConverter.OsPackageName basePackage = osPackageNames.first() ?: emptyBasePackage
+
+    String baseName = bakeRequest.ami_name ?: basePackage.name
+    String release = bakeRequest.ami_suffix ?: timestamp
+    String arch = basePackage.arch ?: "all" // TODO This is defaulted to 'all' since all tests assume this. Skip?
+
+    [ baseName, arch, release, bakeRequest.base_os].findAll{it != "" }.join("-")
```

Comment:
Since the test you're using here is `it != ""`, this leaves us open to concatenating `null` when one of the items in the list is explicitly `null`. While this looks impossible with the current path to this line, I think it would be more future-proof if the test were actually just `it`. Then the result would be false for both an empty string and a `null`.